src/ucp/wireup/select.c

```diff
@@ -225,6 +225,26 @@ static int ucp_wireup_check_amo_flags(const uct_tl_resource_desc_t *resource,
     return 0;
 }
 
+static int
+ucp_wireup_check_keepalive(const ucp_wireup_select_params_t *select_params,
+                           const uct_tl_resource_desc_t *resource,
+                           uint64_t flags, uint64_t required_flags,
+                           const char *title, const char **flag_descs,
+                           char *reason, size_t max)
+{
+    ucp_worker_h worker = select_params->ep->worker;
+    return /* if error handling and keepalive were requested, UCT iface has to
+            * support peer failure (i.e. UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE)
+            * and either built-in keepalive (i.e. UCT_IFACE_FLAG_EP_KEEPALIVE)
+            * or EP checking (i.e. UCT_IFACE_FLAG_EP_CHECK) */
+           !ucp_worker_keepalive_is_enabled(worker) ||
+           !(select_params->ep_init_flags & UCP_EP_INIT_ERR_MODE_PEER_FAILURE) ||
+           ucp_wireup_check_flags(resource, flags, UCT_IFACE_FLAG_EP_KEEPALIVE, title,
+                                  ucp_wireup_iface_flags, reason, max) ||
+           ucp_wireup_check_flags(resource, flags, UCT_IFACE_FLAG_EP_CHECK, title,
+                                  ucp_wireup_iface_flags, reason, max);
```

Comment:
> does it mean that every transport we select must support keepalive or ep_check?

if KA is enabled (now it is by default) and error handling requested, every selected transport has to support built-in KA or ep_check

> it means that before this PR, we could have a case where the user requested err handling mode, but keepalive may not be working because selected transport did not support?

yes, this was a case, e.g. for UD before EP_CHECK was introduced